apps/src/netsim/NetSimRouterNode.js

```diff
@@ -805,64 +970,53 @@ NetSimRouterNode.prototype.canHandleMessage_ = function (message) {
  */
 NetSimRouterNode.prototype.onMessageTableChange_ = function (rows) {
   if (!this.simulateForSender_) {
-    // Not configured to handle anything yet; don't process messages.
+    // Not a simulating router, we can ignore these.
+    // Really should never hit this, since we don't hook up listeners
+    // unless simulating.
     return;
   }
 
-  if (this.isProcessingMessages_) {
-    // We're already in this method, getting called recursively because
-    // we are making changes to the table.  Ignore this call.
-    return;
-  }
+  this.updateRouterQueue_(rows);
 
-  var messages = rows
-      .map(function (row) {
-        return new NetSimMessage(this.shard_, row);
-      }.bind(this))
-      .filter(this.canHandleMessage_.bind(this));
-
-  if (messages.length === 0) {
-    // No messages for us, no work to do.
-    return;
+  if (this.dnsMode === DnsMode.AUTOMATIC) {
+    this.updateAutoDnsQueue_(rows);
   }
+};
 
-  // Setup (sync): Set processing flag
-  logger.info("Router received " + messages.length + " messages");
-  this.isProcessingMessages_ = true;
 
-  // Step 1 (async): Pull all our messages out of storage.
-  NetSimEntity.destroyEntities(messages, function (err) {
-    if (err) {
-      logger.error("Error pulling message off the wire for routing; " + err.message);
-      this.isProcessingMessages_ = false;
-      return;
-    }
+/**
+ * Updates our cache of all messages that are going to the router (regardless
+ * of which simulation will handle them), so we can use it for stats and rate
+ * limiting.
+ * @param {messageRow[]} rows - message table rows
+ */
+NetSimRouterNode.prototype.updateRouterQueue_ = function (rows) {
+  var newQueue = rows.filter(this.isMessageToRouter_.bind(this));
+  if (!_.isEqual(this.routerQueueCache_, newQueue)) {
```

Comment:
Had to look this method up. Surprised they don't call it something like isDeepEqual. Neat that they have it though.